### ISSUE TITLE
RuntimeException thrown while IScheduledExecutorService#scheduleAtFixedRate() breaks following execution

### ISSUE THREAD

**Issue by akvone** [hazelcast/hazelcast#14142]
It's seems to me that any RuntimeException thrown in run() method for IScheduledExecutorService#scheduleAtFixedRate() should not break the re-partitioning logic. 
However the following code proves the opposite.

Run two instances and after some time drop the one which executes task.
* If 'throwException' is set to 'false', then task continues execution after re-partitioning on alive pod;
* If 'throwException' is set to 'true', then task won't be executed anymore. Also adding new HZ member throws DuplicateTaskException, so there are no way to execute task again;

Tested on HZ version 3.9.3 and 3.11, Java 11

```
public class HZTest2 {

  public static final String SERVICE_NAME = "service_name";

  public static void main(String[] args) throws InterruptedException {
    HazelcastInstance hz = Hazelcast.newHazelcastInstance();
    IScheduledExecutorService a = hz.getScheduledExecutorService(SERVICE_NAME);

    try {
      a.scheduleAtFixedRate(new R(), 1, 1, TimeUnit.SECONDS);
    } catch (DuplicateTaskException e){
      System.out.println("Duplicated task");
    }


    Thread.sleep(1_000_000);
  }


  public static class R implements NamedTask, Serializable, Runnable {

    static boolean throwException = false;

    @Override
    public String getName() {
      return "name";
    }

    @Override
    public void run() {
      System.out.println(System.currentTimeMillis()/1000);

      if (throwException){
        throw new RuntimeException();
      }
    }
  }
}

```

Maybe it's an expected behavior, but I actually didn't find any related information for this case.

**Comment by akvone** [comment]
I have a production case related to this topic, so I can describe it if needed

**Comment by tkountis** [comment]
Hi @akvone thanks for your report.
Interestingly enough the correct contract should be the one you report as incorrect :)
When a task encounters an exception all further invocations are suppressed, which is in-line with `j.u.c.ScheduledExecutor` 

```
   /**
     * Creates and executes a periodic action that becomes enabled first
     * after the given initial delay, and subsequently with the given
     * period; that is executions will commence after
     * {@code initialDelay} then {@code initialDelay+period}, then
     * {@code initialDelay + 2 * period}, and so on.
     * If any execution of the task
     * encounters an exception, subsequent executions are suppressed.
     * Otherwise, the task will only terminate via cancellation or
     * termination of the executor.  If any execution of this task
     * takes longer than its period, then subsequent executions
     * may start late, but will not concurrently execute.
     *
     * @param command the task to execute
     * @param initialDelay the time to delay first execution
     * @param period the period between successive executions
     * @param unit the time unit of the initialDelay and period parameters
     * @return a ScheduledFuture representing pending completion of
     *         the task, and whose {@code get()} method will throw an
     *         exception upon cancellation
     * @throws RejectedExecutionException if the task cannot be
     *         scheduled for execution
     * @throws NullPointerException if command is null
     * @throws IllegalArgumentException if period less than or equal to zero
     */
    public ScheduledFuture<?> scheduleAtFixedRate(Runnable command,
                                                  long initialDelay,
                                                  long period,
                                                  TimeUnit unit);```

So what happens, in the even of task migration is *correct*. If you want to have the task continue, then you will have to deal with the exception internally with a `try-catch` approach. Any exception that escapes your `run` will be treated as a task result. `scheduleAtFixedRate` invocation return a `Future` object which can be used to access the status of the task, and `dispose` the task altogether if you want to re-schedule it at any point in the future. 

Now, correct behaviour aside, when the task doesn't get promoted or migrate (i.e no node goes down/added) the task continues execution (even after the exception) on that original node. This is incorrect, i will have a look at that, in order have a consistent behaviour.

**Comment by akvone** [comment]
@tkountis Thank you for your response!
Yes now I see. Repeating stack trace confused me.
Could you please think about adding some notes to documentation. For example:
* line to `IScheduledExecutorService#scheduleAtFixedRate()` javadoc:
```If any execution of the task an exception, subsequent executions are suppressed.```
* something about `try-catch` approach to Hazelcast IMDG Reference Manual.

We have discussed this case within our team. As this fix will be implemented only in 3.12 other developers on previous versions may base their solution on assumption of non-stopping exception handling by Hazelcast itself. 

Thanks.

**Comment by tkountis** [comment]
Thanks for your suggestions @akvone. I will indeed update the doc & manual.
Regarding the fix, it's targeting 3.12 but it will be back-ported to maintenance too.
